src/measure_theory/covering/besicovitch.lean

```diff
@@ -0,0 +1,440 @@
+(r : fin N.succ → ℝ )
+(rpos : ∀ i, 0 < r i)
+(h : ∀ i j, i ≠ j → (r i ≤ dist (c i) (c j) ∧ r j ≤ τ * r i) ∨
+                    (r j ≤ dist (c j) (c i) ∧ r i ≤ τ * r j))
+(hlast : ∀ i < last N, r i ≤ dist (c i) (c (last N)) ∧ r (last N) ≤ τ * r i)
+(inter : ∀ i < last N, dist (c i) (c (last N)) ≤ r i + r (last N))
+
+/-- A metric space has the Besicovitch covering property if there exist `N` and `τ > 1` such that
+there are no satellite configuration of parameter `τ` with `N+1` points. This is the condition that
+guarantees that the Besicovitch covering theorem holds. It is satified by finite-dimensional
+real vector spaces. -/
+class has_besicovitch_covering (α : Type*) [metric_space α] : Prop :=
+(no_satellite_config : ∃ (N : ℕ) (τ : ℝ), 1 < τ ∧ is_empty (besicovitch.satellite_config α N τ))
+
+namespace besicovitch
+
+namespace satellite_config
+variables {α : Type*} [metric_space α] {N : ℕ} {τ : ℝ} (a : satellite_config α N τ)
+
+lemma inter' (i : fin N.succ) : dist (a.c i) (a.c (last N)) ≤ a.r i + a.r (last N) :=
+begin
+  rcases lt_or_le i (last N) with H|H,
+  { exact a.inter i H },
+  { have I : i = last N := top_le_iff.1 H,
+    have := (a.rpos (last N)).le,
+    simp only [I, add_nonneg this this, dist_self] }
+end
+
+lemma hlast' (i : fin N.succ) (h : 1 ≤ τ) : a.r (last N) ≤ τ * a.r i :=
+begin
+  rcases lt_or_le i (last N) with H|H,
+  { exact (a.hlast i H).2 },
+  { have : i = last N := top_le_iff.1 H,
+    rw this,
+    exact le_mul_of_one_le_left (a.rpos _).le h }
+end
+
+end satellite_config
+
+/-! ### Extracting disjoints subfamilies from a ball covering -/
+
+/-- A ball package is a family of balls in a metric space with positive bounded radii. -/
+structure ball_package (β : Type*) (α : Type*) :=
+(c : β → α)
+(r : β → ℝ)
+(rpos : ∀ b, 0 < r b)
+(r_bound : ℝ)
+(r_le : ∀ b, r b ≤ r_bound)
+
+/-- The ball package made of unit balls. -/
+def unit_ball_package (α : Type*) : ball_package α α :=
+{ c := id,
+  r := λ _, 1,
+  rpos := λ _, zero_lt_one,
+  r_bound := 1,
+  r_le := λ _, le_rfl }
+
+instance (α : Type*) : inhabited (ball_package α α) :=
+⟨unit_ball_package α⟩
+
+/-- A Besicovitch tau-package is a family of balls in a metric space with positive bounded radii,
+together with enough data to proceed with the Besicovitch greedy algorithm. We register this in
+a single structure to make sure that all our constructions in this algorithm only depend on
+one variable. -/
+structure tau_package (β : Type*) (α : Type*) extends ball_package β α :=
+(τ : ℝ)
+(one_lt_tau : 1 < τ)
+
+instance (α : Type*) : inhabited (tau_package α α) :=
+⟨{ τ := 2,
+  one_lt_tau := one_lt_two,
+  .. unit_ball_package α }⟩
+
+variables {α : Type*} [metric_space α] {β : Type u}
+
+namespace tau_package
+
+variables [nonempty β] (p : tau_package β α)
+include p
+
+/-- Choose inductively large balls with centers that are not contained in the union of already
+chosen balls. This is a transfinite induction. -/
+noncomputable def f : ordinal.{u} → β
```

Comment:
Can this have a name that indicates it is choosing something? Maybe just `choose`?

Reply:
I went for `index`.